Source/Core/Core/PowerPC/JitInterface.cpp

```diff
@@ -123,58 +113,97 @@ void JitInterface::UpdateMembase()
+    });
+  }
+  else if (m_jit->IsProfilingEnabled())
+  {
+    u64 overall_cycles_spent = 0;
+    JitBlock::ProfileData::Clock::duration overall_time_spent = {};
+    m_jit->GetBlockCache()->RunOnBlocks(guard, [&](const JitBlock& block) {
+      overall_cycles_spent += block.cycle_count * block.profile_data->run_count;
+      overall_time_spent += block.profile_data->time_spent;
+    });
+    m_jit->GetBlockCache()->RunOnBlocks(guard, [&](const JitBlock& block) {
+      const Common::Symbol* symbol =
+          m_jit->m_ppc_symbol_db.GetSymbolFromAddr(block.effectiveAddress);
+      const JitBlock::ProfileData* const data = block.profile_data.get();
+
+      const auto cycles_spent = block.cycle_count * data->run_count;
+      const auto time_spent =
+          std::chrono::duration_cast<std::chrono::duration<double, std::micro>>(data->time_spent);
+      // ProfileData can be wiped, leaving the potential for division by zero.
+      const auto cycles_percent =
+          overall_cycles_spent == 0 ? 0.0 : 100.0 * cycles_spent / overall_cycles_spent;
+      const auto time_percent = overall_time_spent == JitBlock::ProfileData::Clock::duration{0} ?
+                                    0.0 :
+                                    100.0 * data->time_spent.count() / overall_time_spent.count();
+      const auto time_average =
+          data->run_count == 0 ?
+              std::chrono::duration<double, std::nano>{0} :
+              std::chrono::duration_cast<std::chrono::duration<double, std::nano>>(
+                  data->time_spent) /
+                  data->run_count;
+
+      const auto host_near_code_size = block.near_end - block.near_begin;
+      const auto host_far_code_size = block.far_end - block.far_begin;
+
+      fmt::println(file,
+                   "{}\t{:08x}\t{}\t{}\t{}\t{}\t{}\t{}\t{:.6f}\t{:.6f}\t{:.6f}\t{:.6f}\t\"{}\"",
+                   GetDescription(block.feature_flags), block.effectiveAddress,
+                   block.physical_addresses.size() * sizeof(UGeckoInstruction), block.cycle_count,
+                   host_near_code_size, host_far_code_size, data->run_count, cycles_spent,
+                   cycles_percent, time_spent.count(), time_average.count(), time_percent,
+                   symbol ? std::string_view{symbol->name} : "");
+    });
+  }
+  else
+  {
+    m_jit->GetBlockCache()->RunOnBlocks(guard, [&](const JitBlock& block) {
+      const Common::Symbol* symbol =
+          m_jit->m_ppc_symbol_db.GetSymbolFromAddr(block.effectiveAddress);
+
+      const auto host_near_code_size = block.near_end - block.near_begin;
+      const auto host_far_code_size = block.far_end - block.far_begin;
+
+      fmt::println(file, "{}\t{:08x}\t{}\t{}\t{}\t{}\t-\t-\t-\t-\t-\t-\t\"{}\"",
+                   GetDescription(block.feature_flags), block.effectiveAddress,
+                   block.physical_addresses.size() * sizeof(UGeckoInstruction), block.cycle_count,
```

Comment:
`physical_addresses.size() * sizeof(UGeckoInstruction)` isn't an entirely accurate way of determining the number of emulated instructions if branch following has inlined the same code twice. You should use `block.originalSize * sizeof(UGeckoInstruction)` instead.

Reply:
I need your help to understand: why is `JitBlockData::originalSize` important? It is the number of instructions in a `PPCAnalyst::CodeBuffer` that no longer exists. Yes, it can tell you that PPC instructions were recompiled multiple times, but not which ones or how many times. The only code currently using the `originalSize` data member (`JitInterface::GetHostCode`) is *misusing* it by assuming that the instruction coverage of a JitBlock is contiguous and non-repeating from the effective address. The `GetHostCode` function will be removed by my upcoming JitWidget refresh, so I have avoided using the `originalSize` data member in any new places in order to remove this data member. The only way this information could be useful, in my opinion, would be if the `CodeBuffer` were preserved in the JitBlock (a `std::move` seems the obvious choice). Then, my JITWidget could present a disassembly of *that* rather than reconstructing the instruction coverage from the `physical_addresses` set like it currently has to.

So tell me what to do. Is it worth the memory footprint to keep around a bunch of 2400% blown-up PPC Code buffers for obscure debugging features? Am I right that the size of the missing buffer is irrelevant? Should I suck it up and just do what you said?

Reply:
I also have a few ideas for making the `PPCAnalyst::CodeOp` struct smaller using a union. It seems like there are data members in it that are exclusive to certain instruction families.

Reply:
I have thought about it further. The answer seems to be continuing to use `JitBlockData::originalSize` as you have requested, then for the JitWidget refresh I will convince you that it is worth moving the code buffer into the JitBlock to replace the `originalSize` member. Sounds good?

Reply:
> I need your help to understand: why is `JitBlockData::originalSize` important? It is the number of instructions in a `PPCAnalyst::CodeBuffer` that no longer exists.

The key question here is: What are we trying to print to the file? You probably know your reasoning for including this `ppcSize` column better than me, but the kinds of uses I imagine for it are things along the lines of calculating the code size blowup, i.e. the size of the host code divided by the size of the guest code, which is used as a kind of measurement of how efficient the JIT is at translating particular blocks.

Imagine that we have the following three functions:

`f`: 1000 bytes of PPC code.
`g`: A call to `f`, followed by a BLR. This is in itself 8 bytes of PPC code.
`h`: Two calls to `f`, followed by a BLR. This is in itself 12 bytes of PPC code.

Let's say that `f` translates into 1600 bytes of AArch64 code. That means the blowup for `f` is 1.6.

Now let's try `g`. Because we have branch following, the code of `f` becomes inlined when we JIT `g`. This means that `ppcSize` is 1008 bytes, and the size of the translated AArch64 code is a little above 1600 bytes (I don't recall how big the emitted code for BL and BLR would be), for a blowup of around 1.6 once again.

Finally, `h`. Because `f` gets inlined twice here, the AArch64 code will be somewhere around 3200 bytes. But what will `ppcSize` be? If we calculate it as `block.originalSize * sizeof(UGeckoInstruction)`, it will be **2012 bytes**. If we calculate it as `physical_addresses.size() * sizeof(UGeckoInstruction)`, it will be **1012 bytes**. For the former the blowup is still somewhere around 1.6, but for the latter it has ballooned to around 3.2. Why? Because we're miscounting the PPC instructions. We did JIT each instruction of `f` twice, so not counting each instruction of `f` twice when calculating the `ppcSize` value is in my opinion a bug.

> Is it worth the memory footprint to keep around a bunch of 2400% blown-up PPC Code buffers for obscure debugging features?

That does not sound especially worth it. We definitely shouldn't do it when debugging is disabled, at least.

> Am I right that the size of the missing buffer is irrelevant?

The buffer itself is of no importance to me for this. But it so happens that the size of this buffer matches the number of instructions we actually jitted.

> Edit: All this talk, and it finally occurred to me the reason for the PPCAnalyst in the current JITWidget's design is to avoid this data sticking around in memory. Alright, I believe I understand what must be done now.

I'm not sure if I understand :) But this PR is now using originalSize, so I have no remaining review complaints about this. We can always discuss this more outside the scope of this PR.